webhooks/virtualmachine/mutation/scsi_controller.go

```diff
@@ -0,0 +1,306 @@
+		// the first available slot on the lowest bus number of the
+		// matching controller. If we do not specify the busNumber in
+		// the volume explicitly, it is possible that by the time we
+		// reconcile this volume, another slot has opened up on an
+		// existing controller. We just ended up adding a controller
+		// that does not have any devices attached to it. Let's try to
+		// avoid that.
+		if pvc.ControllerBusNumber == nil {
+			busNum := targetController.BusNumber
+			vol.PersistentVolumeClaim.ControllerBusNumber = &busNum
+			updatedVM = true
+		}
+
+		// Set controllerType to SCSI if not already set
+		if pvc.ControllerType == "" {
+			vol.PersistentVolumeClaim.ControllerType = vmopv1.VirtualControllerTypeSCSI
+			updatedVM = true
+		}
+	}
+
+	return updatedVM, nil
+}
+
+// determineTargetController determines which controller should be
+// used for a volume and whether a new controller needs to be created.
+func determineTargetController(
+	pvc *vmopv1.PersistentVolumeClaimVolumeSource,
+	existingControllers map[int32]*vmopv1.SCSIControllerSpec,
+	statusDeviceCounts, volumeAssignments map[int32]int32,
+) (*vmopv1.SCSIControllerSpec, bool) {
+
+	// We only handle SCSI controllers here.
+	// TODO: handle other controllers in a followup.
+	if pvc.ControllerType != "" && pvc.ControllerType != vmopv1.VirtualControllerTypeSCSI {
+		return nil, false
+	}
+
+	// If a specific controller is requested, check if it already
+	// exists. If not, we need to create it.
+	if pvc.ControllerBusNumber != nil {
+		busNum := *pvc.ControllerBusNumber
+
+		if controller, exists := existingControllers[busNum]; exists {
+			return controller, false
+		}
+
+		// Need to create this specific controller
+		return createSCSIController(busNum, vmopv1.VirtualControllerSharingModeNone), true
+	}
+
+	// Determine controller requirements based on volume configuration
+	// After SetPVCVolumeDefaultsOnCreate has run, the volume will
+	// have diskMode and sharingMode set
+
+	// Check if this is an OracleRAC volume (either by applicationType
+	// or by explicit sharingMode=MultiWriter). OracleRAC volumes (with
+	// sharingMode=MultiWriter) need a controller with
+	// sharingMode=None
+	if pvc.ApplicationType == vmopv1.VolumeApplicationTypeOracleRAC ||
+		pvc.SharingMode == vmopv1.VolumeSharingModeMultiWriter {
+		return findOrCreateControllerWithSharing(
+			vmopv1.VirtualControllerSharingModeNone,
+			existingControllers,
+			statusDeviceCounts,
+			volumeAssignments,
+		)
+	}
+
+	// Check if this is a MicrosoftWSFC volume
+	// WSFC volumes need a controller with sharingMode=Physical
+	if pvc.ApplicationType == vmopv1.VolumeApplicationTypeMicrosoftWSFC {
+		return findOrCreateControllerWithSharing(
+			vmopv1.VirtualControllerSharingModePhysical,
+			existingControllers,
+			statusDeviceCounts,
+			volumeAssignments,
+		)
+	}
+
+	// Default behavior: find first SCSI controller with available
+	// slots If no controller with available slots exists, create a
```

Comment:
Okay, now I think you are just allergic to punctuation. I get it above, where you did not end a sentence with a period because the next one began on a new line. There you had a shred of an excuse. In this case you literally start a new sentence on the same line with no indication other than the capitalization of `If`. 🤣